[Traceable FSDP2][Dynamo] Support reconstructing CUDA event object within Dynamo graph

yf225 opened the pull request:
`torch.cuda.Event` objects are different from `torch.cuda.Stream` in that events are not pooled, meaning we can't look up a previously created CUDA event object by ID. This prevents CUDA event object created outside of the Dynamo graph from being used within the graph (since Dynamo needs a way to emit a `call_function` line in the graph that does the retrieval of the event object for downstream op use). This PR adds a simple object pool within Dynamo utility, to support looking up CUDA event object by ID from within the Dynamo graph.

After this PR, if a user creates a CUDA event object outside of the graph and use that event within the graph, the behavior will exactly match eager.

Test commands:
- `pytest -rA test/dynamo/test_ctx_manager.py::CtxManagerTests::test_cuda_event_created_outside_of_graph`
- `pytest -rA test/dynamo/test_ctx_manager.py::CtxManagerTests::test_cuda_event_across_graph_break`

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #133723
* #133657
* __->__ #133635
* #133636
* #133531
* #133532



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

yf225 committed:
[Dynamo] Ignore calling .wait_event() on CUDA events created outside of the current graph

[ghstack-poisoned]

yf225 committed:
Update on "[Traceable FSDP2][Dynamo] Ignore calling .wait_event() on CUDA events created outside of the current graph"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

yifuwang commented on torch/_dynamo/variables/ctx_manager.py:
```diff
@@ -958,11 +958,22 @@ def call_method(
+        elif name == "wait_event":
+            # NOTE: Unlike CUDA streams, currently there is no way to look up a previously-created CUDA event by id
+            # (i.e. CUDA events are not cached in a pool), therefore there is no way to create a proxy for it and
+            # reconstruct it within the graph. As a result, Dynamo graph does not support calling `.wait_event()`
```
I might be missing something, but is the correctness still guaranteed? (e.g. all event has been waited prior to this point)

yf225 replied:
I believe events are not waited even before this PR, so functionally it still behaves the same.

yf225 replied:
also prior to this PR, if the CUDA event was created outside of the graph, we are issuing `.wait_event(None)` which throws error. With changes in this PR, we will not issue that `.wait_event(None)` line, thus eliminating the error.

yf225 committed:
Update on "[Traceable FSDP2][Dynamo] Ignore calling .wait_event() on CUDA events created outside of the current graph"

- In general, regardless of where the CUDA event is created, they are not waited under Dynamo.
- For CUDA event created outside of current graph, Dynamo will issue `stream.wait_event(None)` in the Dynamo graph, which just fails during execution (we can't wait on a None event).

This PR addresses the second case, by not issuing `stream.wait_event(None)` thus allowing execution to continue.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

yf225 committed:
Update on "[Traceable FSDP2][Dynamo] Ignore calling .wait_event() on CUDA events created outside of the current graph"

- In general, regardless of where the CUDA event is created, they are not waited under Dynamo.
- For CUDA event created outside of current graph, Dynamo will issue `stream.wait_event(None)` in the Dynamo graph, which just fails during execution (we can't wait on a None event).

This PR addresses the second case, by not issuing `stream.wait_event(None)` thus allowing execution to continue.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

yf225 committed:
Update on "[Traceable FSDP2][Dynamo] Ignore calling .wait_event() on CUDA events created outside of the current graph"

- In general, regardless of where the CUDA event is created, they are not waited under Dynamo.
- For CUDA event created outside of current graph, Dynamo will issue `stream.wait_event(None)` in the Dynamo graph, which just fails during execution (we can't wait on a None event).

This PR addresses the second case, by not issuing `stream.wait_event(None)` thus allowing execution to continue.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

yf225 committed:
Update on "[Traceable FSDP2][Dynamo] Support reconstructing CUDA event object within Dynamo graph"

`torch.cuda.Event` objects are different from `torch.cuda.Stream` in that events are not pooled, meaning we can't look up a previously created CUDA event object by ID. This PR adds a simple object pool within Dynamo utility, to support looking up CUDA event object by ID from within the Dynamo graph.

After this PR, if a user creates a CUDA event object outside of the graph and use that event within the graph, the behavior will exactly match eager.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

yf225 committed:
Update on "[Traceable FSDP2][Dynamo] Support reconstructing CUDA event object within Dynamo graph"

`torch.cuda.Event` objects are different from `torch.cuda.Stream` in that events are not pooled, meaning we can't look up a previously created CUDA event object by ID. This PR adds a simple object pool within Dynamo utility, to support looking up CUDA event object by ID from within the Dynamo graph.

After this PR, if a user creates a CUDA event object outside of the graph and use that event within the graph, the behavior will exactly match eager.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

yf225 committed:
Update on "[Traceable FSDP2][Dynamo] Support reconstructing CUDA event object within Dynamo graph"

`torch.cuda.Event` objects are different from `torch.cuda.Stream` in that events are not pooled, meaning we can't look up a previously created CUDA event object by ID. This PR adds a simple object pool within Dynamo utility, to support looking up CUDA event object by ID from within the Dynamo graph.

After this PR, if a user creates a CUDA event object outside of the graph and use that event within the graph, the behavior will exactly match eager.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

yifuwang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -799,8 +799,16 @@ def build_key_value(i, k, v):
             return SDPAParamsVariable.create(self.tx, value, self.source)
         elif isinstance(value, _EventBase):
             self.install_guards(GuardBuilder.ID_MATCH)
+            torch._dynamo.utils.store_user_object(value)
```
Consider reflecting "weakref" in the name of this util function so it's clear that dynamo is not keeping them alive.

yf225 committed:
Update on "[Traceable FSDP2][Dynamo] Support reconstructing CUDA event object within Dynamo graph"

`torch.cuda.Event` objects are different from `torch.cuda.Stream` in that events are not pooled, meaning we can't look up a previously created CUDA event object by ID. This PR adds a simple object pool within Dynamo utility, to support looking up CUDA event object by ID from within the Dynamo graph.

After this PR, if a user creates a CUDA event object outside of the graph and use that event within the graph, the behavior will exactly match eager.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

yf225 committed:
Update on "[Traceable FSDP2][Dynamo] Support reconstructing CUDA event object within Dynamo graph"

`torch.cuda.Event` objects are different from `torch.cuda.Stream` in that events are not pooled, meaning we can't look up a previously created CUDA event object by ID. This PR adds a simple object pool within Dynamo utility, to support looking up CUDA event object by ID from within the Dynamo graph.

After this PR, if a user creates a CUDA event object outside of the graph and use that event within the graph, the behavior will exactly match eager.

Test commands:
- `pytest -rA test/dynamo/test_ctx_manager.py::CtxManagerTests::test_cuda_event_created_outside_of_graph`
- `pytest -rA test/dynamo/test_ctx_manager.py::CtxManagerTests::test_cuda_event_across_graph_break`




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

yf225 committed:
Update on "[Traceable FSDP2][Dynamo] Support reconstructing CUDA event object within Dynamo graph"

`torch.cuda.Event` objects are different from `torch.cuda.Stream` in that events are not pooled, meaning we can't look up a previously created CUDA event object by ID. This PR adds a simple object pool within Dynamo utility, to support looking up CUDA event object by ID from within the Dynamo graph.

After this PR, if a user creates a CUDA event object outside of the graph and use that event within the graph, the behavior will exactly match eager.

Test commands:
- `pytest -rA test/dynamo/test_ctx_manager.py::CtxManagerTests::test_cuda_event_created_outside_of_graph`
- `pytest -rA test/dynamo/test_ctx_manager.py::CtxManagerTests::test_cuda_event_across_graph_break`




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

yf225 committed:
Update on "[Traceable FSDP2][Dynamo] Support reconstructing CUDA event object within Dynamo graph"

`torch.cuda.Event` objects are different from `torch.cuda.Stream` in that events are not pooled, meaning we can't look up a previously created CUDA event object by ID. This PR adds a simple object pool within Dynamo utility, to support looking up CUDA event object by ID from within the Dynamo graph.

After this PR, if a user creates a CUDA event object outside of the graph and use that event within the graph, the behavior will exactly match eager.

Test commands:
- `pytest -rA test/dynamo/test_ctx_manager.py::CtxManagerTests::test_cuda_event_created_outside_of_graph`
- `pytest -rA test/dynamo/test_ctx_manager.py::CtxManagerTests::test_cuda_event_across_graph_break`




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

yf225 commented:
@pytorchbot merge
